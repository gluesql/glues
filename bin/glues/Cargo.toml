[package]
name = "glues"
version.workspace = true
edition.workspace = true
license.workspace = true
repository.workspace = true
authors.workspace = true
description = "Unified CLI entry point for the Glues toolkit"

[dependencies]
clap = { version = "4.5.4", features = ["derive"] }
color-eyre = "0.6.3"
glues-tui = { path = "../../tui", package = "glues-tui", default-features = false, features = ["cli"] }
glues-server = { path = "../../server" }
tokio = { version = "1.41.0", features = ["macros", "rt-multi-thread"] }

[features]
default = []
