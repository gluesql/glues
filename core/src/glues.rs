use {
    crate::{
        Event, Result, Transition,
        backend::CoreBackend,
        state::{EntryState, State},
        task::{Task, handle_tasks},
    },
    std::{
        collections::VecDeque,
        sync::{
            Arc, Mutex,
            mpsc::{Sender, channel},
        },
        thread::JoinHandle,
    },
};

pub struct Glues {
    pub db: Option<Box<dyn CoreBackend + Send>>,
    pub state: State,

    pub task_tx: Sender<Task>,
    pub task_handle: JoinHandle<()>,
    pub transition_queue: Arc<Mutex<VecDeque<Transition>>>,
}

impl Default for Glues {
    fn default() -> Self {
        Self::new()
    }
}

impl Glues {
    pub fn new() -> Self {
        let transition_queue = Arc::new(Mutex::new(VecDeque::new()));
        let (task_tx, task_rx) = channel();
        let task_handle = handle_tasks(task_rx, &transition_queue);

        Self {
            db: None,
            state: EntryState.into(),
            task_tx,
            task_handle,
            transition_queue,
        }
    }

    pub async fn dispatch(&mut self, event: Event) -> Result<Transition> {
        State::consume(self, event).await
    }
}
