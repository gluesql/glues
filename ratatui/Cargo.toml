[package]
name = "glues"
authors.workspace = true
version.workspace = true
edition.workspace = true
license.workspace = true
repository.workspace = true
description = "Lightweight TUI note-taking app with Git, CSV, and JSON support - privacy-focused and sync-enabled"
readme = "../README.md"
keywords = ["tui", "note-taking", "notes-app", "ratatui", "data-privacy"]
default-run = "glues"

[dependencies]
glues-core.workspace = true
gluesql.workspace = true
futures = "0.3.30"
async-io = "2.3.2"
ratatui = "0.28.1"
color-eyre = "0.6.3"
tui-big-text = "0.6.0"
tui-textarea = "0.6.1"
home = "0.5.9"
