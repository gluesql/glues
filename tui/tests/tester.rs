use color_eyre::Result;
use glues::{App, config, logger};
use ratatui::{
    Terminal,
    backend::TestBackend,
    crossterm::event::{
        Event as Input, KeyCode, KeyEvent as CKeyEvent, KeyEventKind, KeyModifiers,
    },
};

pub struct Tester {
    pub app: App,
    pub term: Terminal<TestBackend>,
}

// Bring a concise snapshot macro into test files.
// Usage: snap!(t, "name");
#[allow(unused_macros)]
macro_rules! snap {
    ($t:expr, $name:expr) => {{
        let text = $t.snapshot_text();
        insta::assert_snapshot!($name, text);
    }};
}

impl Tester {
    pub async fn new() -> Result<Self> {
        let cwd = std::env::current_dir()?;
        std::fs::create_dir_all(cwd.join(".glues"))?;
        unsafe {
            std::env::set_var("HOME", &cwd);
        }
        config::init().await;
        logger::init().await;

        let backend = TestBackend::new(120, 40);
        let term = Terminal::new(backend)?;
        let app = App::new();
        Ok(Self { app, term })
    }

    pub fn draw(&mut self) -> color_eyre::Result<()> {
        self.term.draw(|f| self.app.draw(f))?;
        Ok(())
    }

    pub async fn press(&mut self, c: char) -> bool {
        self.handle_input(Input::Key(CKeyEvent::new(
            KeyCode::Char(c),
            KeyModifiers::NONE,
        )))
        .await
    }

    #[allow(dead_code)]
    pub async fn ctrl(&mut self, c: char) -> bool {
        self.handle_input(Input::Key(CKeyEvent::new(
            KeyCode::Char(c),
            KeyModifiers::CONTROL,
        )))
        .await
    }

    #[allow(dead_code)]
    pub async fn key(&mut self, code: KeyCode) -> bool {
        self.handle_input(Input::Key(CKeyEvent::new(code, KeyModifiers::NONE)))
            .await
    }

    #[allow(dead_code)]
    pub async fn open_instant(&mut self) -> Result<()> {
        self.draw()?;
        let _ = self.press('1').await;
        self.draw()?;
        Ok(())
    }

    #[allow(dead_code)]
    pub fn snapshot_text(&self) -> String {
        buffer_lines(&self.term).join("\n")
    }

    async fn handle_input(&mut self, input: Input) -> bool {
        if !matches!(
            input,
            Input::Key(CKeyEvent {
                kind: KeyEventKind::Press,
                ..
            })
        ) {
            return false;
        }

        match input {
            Input::Key(CKeyEvent {
                code: KeyCode::Char('c'),
                modifiers: KeyModifiers::CONTROL,
                ..
            }) => true,
            _ => {
                let action = self.app.context_mut().consume(&input).await;
                self.app.handle_action(action, input).await
            }
        }
    }
}

fn buffer_lines(term: &Terminal<TestBackend>) -> Vec<String> {
    let buf = term.backend().buffer().clone();
    let area = buf.area();
    let mut lines = Vec::with_capacity(area.height as usize);
    for y in 0..area.height {
        let mut line = String::new();
        for x in 0..area.width {
            line.push_str(buf[(x, y)].symbol());
        }
        lines.push(line);
    }
    lines
}
