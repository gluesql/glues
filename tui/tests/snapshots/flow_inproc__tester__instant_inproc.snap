---
source: tui/tests/tester.rs
assertion_line: 999
expression: lines
snapshot_kind: text
---
[
    " Directory 'Notes' selected                                                                           \u{e0be} [?] Show keymap ",
    "[Browser]                                   ▐[Editor]                                                                   ",
    " \u{f0770} Notes                                    ▐ 1 Welcome to Glues :D                                                     ",
    "   \u{f11d7} Sample Note                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
    "                                            ▐                                                                           ",
]
